File: PSG Officiel/utils/permissions.py
# utils/permissions.py - VERSION AVEC SALONS SANS COINS
import discord
import json
import os
from typing import Optional, List

SERVERS_DIR = "data/servers"

def init_server_config(guild_id: str, guild_name: str):
    """Initialise la configuration d'un serveur"""
    os.makedirs(SERVERS_DIR, exist_ok=True)
    
    config_path = f"{SERVERS_DIR}/{guild_id}.json"
    if not os.path.exists(config_path):
        config = {
            "guild_id": guild_id,
            "guild_name": guild_name,
            "channels": {
                "solde": [],
                "packs": [],
                "collection": []
            },
            "roles": {
                "admin": [],
                "moderator": []
            },
            "no_coins_channels": []  # ✅ Nouvelle section
        }
        with open(config_path, 'w', encoding='utf-8') as f:
            json.dump(config, f, indent=4, ensure_ascii=False)
        return config
    
    return load_server_config(guild_id)

def load_server_config(guild_id: str) -> Optional[dict]:
    """Charge la configuration d'un serveur"""
    config_path = f"{SERVERS_DIR}/{guild_id}.json"
    if os.path.exists(config_path):
        with open(config_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    return None

def save_server_config(guild_id: str, config: dict):
    """Sauvegarde la configuration d'un serveur"""
    os.makedirs(SERVERS_DIR, exist_ok=True)
    config_path = f"{SERVERS_DIR}/{guild_id}.json"
    with open(config_path, 'w', encoding='utf-8') as f:
        json.dump(config, f, indent=4, ensure_ascii=False)

def check_channel_permission(interaction: discord.Interaction, command_name: str) -> bool:
    """Vérifie si la commande peut être exécutée dans ce salon"""
    guild_id = str(interaction.guild_id)
    channel_id = str(interaction.channel_id)
    
    config = load_server_config(guild_id)
    if not config:
        # Si pas de config, autoriser partout par défaut
        return True
    
    allowed_channels = config.get("channels", {}).get(command_name, [])
    
    # Si la liste est vide, autoriser partout
    if not allowed_channels:
        return True
    
    # Convertir tous les IDs en string pour la comparaison
    allowed_channels_str = [str(ch_id) for ch_id in allowed_channels]
    
    # Vérifier si le salon est dans la liste
    return channel_id in allowed_channels_str

def get_allowed_channel(guild_id: int, command_name: str, bot_instance=None) -> Optional[discord.TextChannel]:
    """
    Récupère le premier salon autorisé pour une commande donnée
    
    Args:
        guild_id: ID du serveur Discord (int)
        command_name: Nom de la commande (ex: "packs", "collection", "solde")
        bot_instance: Instance du bot (optionnel, sera importé si None)
    
    Returns:
        discord.TextChannel ou None si aucun salon configuré
    """
    config = load_server_config(str(guild_id))
    
    if not config:
        return None
    
    # Récupérer la liste des channel_ids autorisés pour cette commande
    channel_ids = config.get("channels", {}).get(command_name, [])
    
    if not channel_ids:
        return None
    
    # Utiliser le bot passé en paramètre ou essayer de l'importer
    bot = bot_instance
    if bot is None:
        try:
            import main
            bot = main.bot
        except (ImportError, AttributeError):
            return None
    
    if bot is None:
        return None
    
    # Récupérer le serveur
    guild = bot.get_guild(guild_id)
    if not guild:
        return None
    
    # Retourner le premier salon trouvé
    for channel_id in channel_ids:
        channel = guild.get_channel(int(channel_id))
        if channel and isinstance(channel, discord.TextChannel):
            return channel
    
    return None

def check_role_permission(interaction: discord.Interaction, permission_type: str) -> bool:
    """Vérifie si l'utilisateur a les permissions requises"""
    # TOUJOURS autoriser le propriétaire du bot
    if interaction.user.id == 878724920987766796:
        return True
    
    # Si c'est le propriétaire du serveur, toujours autoriser
    if interaction.user.id == interaction.guild.owner_id:
        return True
    
    guild_id = str(interaction.guild_id)
    config = load_server_config(guild_id)
    
    if not config:
        # Si pas de config, vérifier les permissions Discord natives
        if permission_type == "admin":
            return interaction.user.guild_permissions.administrator
        elif permission_type == "moderator":
            return interaction.user.guild_permissions.moderate_members or \
                   interaction.user.guild_permissions.administrator
    
    allowed_roles = config.get("roles", {}).get(permission_type, [])
    
    # Si la liste est vide, vérifier les permissions Discord natives
    if not allowed_roles:
        if permission_type == "admin":
            return interaction.user.guild_permissions.administrator
        elif permission_type == "moderator":
            return interaction.user.guild_permissions.moderate_members or \
                   interaction.user.guild_permissions.administrator
    
    # Vérifier si l'utilisateur a un des rôles autorisés
    user_role_ids = [str(role.id) for role in interaction.user.roles]
    has_role = any(role_id in allowed_roles for role_id in user_role_ids)
    
    # Toujours autoriser aussi si permissions Discord natives
    if not has_role:
        if permission_type == "admin":
            return interaction.user.guild_permissions.administrator
        elif permission_type == "moderator":
            return interaction.user.guild_permissions.moderate_members or \
                   interaction.user.guild_permissions.administrator
    
    return has_role

def add_channel_permission(guild_id: str, command_name: str, channel_id: str) -> bool:
    """Ajoute un salon autorisé pour une commande"""
    config = load_server_config(guild_id)
    if not config:
        return False
    
    if command_name not in config["channels"]:
        config["channels"][command_name] = []
    
    if channel_id not in config["channels"][command_name]:
        config["channels"][command_name].append(channel_id)
        save_server_config(guild_id, config)
        return True
    
    return False

def remove_channel_permission(guild_id: str, command_name: str, channel_id: str) -> bool:
    """Retire un salon autorisé pour une commande"""
    config = load_server_config(guild_id)
    if not config:
        return False
    
    if command_name in config["channels"] and channel_id in config["channels"][command_name]:
        config["channels"][command_name].remove(channel_id)
        save_server_config(guild_id, config)
        return True
    
    return False

def add_role_permission(guild_id: str, permission_type: str, role_id: str) -> bool:
    """Ajoute un rôle autorisé"""
    config = load_server_config(guild_id)
    if not config:
        return False
    
    if permission_type not in config["roles"]:
        config["roles"][permission_type] = []
    
    if role_id not in config["roles"][permission_type]:
        config["roles"][permission_type].append(role_id)
        save_server_config(guild_id, config)
        return True
    
    return False

def remove_role_permission(guild_id: str, permission_type: str, role_id: str) -> bool:
    """Retire un rôle autorisé"""
    config = load_server_config(guild_id)
    if not config:
        return False
    
    if permission_type in config["roles"] and role_id in config["roles"][permission_type]:
        config["roles"][permission_type].remove(role_id)
        save_server_config(guild_id, config)
        return True
    
    return False

def get_allowed_channels(guild_id: str, command_name: str) -> List[str]:
    """Récupère la liste des salons autorisés pour une commande"""
    config = load_server_config(guild_id)
    if not config:
        return []
    
    return config.get("channels", {}).get(command_name, [])

def get_allowed_roles(guild_id: str, permission_type: str) -> List[str]:
    """Récupère la liste des rôles autorisés pour un type de permission"""
    config = load_server_config(guild_id)
    if not config:
        return []
    
    return config.get("roles", {}).get(permission_type, [])


# ============================================
# ✅ FONCTIONS POUR LES SALONS SANS COINS
# ============================================

def get_no_coins_channels(guild_id: str) -> List[str]:
    """Récupère la liste des salons où les membres ne gagnent pas de coins"""
    config = load_server_config(guild_id)
    if not config:
        return []
    
    return config.get("no_coins_channels", [])


def add_no_coins_channel(guild_id: str, channel_id: str) -> bool:
    """Ajoute un salon à la liste des salons sans coins"""
    config = load_server_config(guild_id)
    if not config:
        return False
    
    if "no_coins_channels" not in config:
        config["no_coins_channels"] = []
    
    if channel_id not in config["no_coins_channels"]:
        config["no_coins_channels"].append(channel_id)
        save_server_config(guild_id, config)
        return True
    
    return False


def remove_no_coins_channel(guild_id: str, channel_id: str) -> bool:
    """Retire un salon de la liste des salons sans coins"""
    config = load_server_config(guild_id)
    if not config:
        return False
    
    if "no_coins_channels" in config and channel_id in config["no_coins_channels"]:
        config["no_coins_channels"].remove(channel_id)
        save_server_config(guild_id, config)
        return True
    
    return False


def is_coins_disabled_channel(guild_id: str, channel_id: str) -> bool:
    """Vérifie si un salon est dans la liste des salons sans coins"""
    no_coins_channels = get_no_coins_channels(guild_id)
    return str(channel_id) in no_coins_channels

File: src/utils/permissions.js
// src/utils/permissions.js - Gestion des permissions par serveur (Enmap)
const {
  initServerConfig,
  loadServerConfig,
  saveServerConfig,
} = require('./database');

const OWNER_ID = '878724920987766796';

// ==================== PERMISSIONS SALONS ====================

function checkChannelPermission(interaction, commandName) {
  const config = loadServerConfig(interaction.guildId);
  if (!config) return true;

  const allowedChannels = config.channels?.[commandName] || [];
  if (!allowedChannels.length) return true;

  return allowedChannels.map(String).includes(String(interaction.channelId));
}

function getAllowedChannel(guildId, commandName, client) {
  const config = loadServerConfig(String(guildId));
  if (!config) return null;

  const channelIds = config.channels?.[commandName] || [];
  if (!channelIds.length) return null;

  const guild = client?.guilds?.cache?.get(String(guildId));
  if (!guild) return null;

  for (const channelId of channelIds) {
    const channel = guild.channels.cache.get(String(channelId));
    if (channel) return channel;
  }
  return null;
}

// ==================== PERMISSIONS RÔLES ====================

function checkRolePermission(interaction, permissionType) {
  if (interaction.user.id === OWNER_ID) return true;
  if (interaction.user.id === interaction.guild?.ownerId) return true;

  const config = loadServerConfig(interaction.guildId);

  const nativeCheck = () => {
    const perms = interaction.member?.permissions;
    if (permissionType === 'admin') return perms?.has('Administrator') ?? false;
    if (permissionType === 'moderator') return perms?.has('ModerateMembers') || perms?.has('Administrator') || false;
    return false;
  };

  if (!config) return nativeCheck();

  const allowedRoles = config.roles?.[permissionType] || [];
  if (!allowedRoles.length) return nativeCheck();

  const userRoleIds = interaction.member?.roles?.cache?.map(r => String(r.id)) || [];
  const hasRole = allowedRoles.some(roleId => userRoleIds.includes(String(roleId)));

  return hasRole || nativeCheck();
}

// ==================== GESTION SALONS ====================

function addChannelPermission(guildId, commandName, channelId) {
  const config = loadServerConfig(guildId);
  if (!config) return false;
  if (!config.channels[commandName]) config.channels[commandName] = [];
  if (!config.channels[commandName].includes(String(channelId))) {
    config.channels[commandName].push(String(channelId));
    saveServerConfig(guildId, config);
    return true;
  }
  return false;
}

function removeChannelPermission(guildId, commandName, channelId) {
  const config = loadServerConfig(guildId);
  if (!config) return false;
  const idx = (config.channels[commandName] || []).indexOf(String(channelId));
  if (idx !== -1) {
    config.channels[commandName].splice(idx, 1);
    saveServerConfig(guildId, config);
    return true;
  }
  return false;
}

function getAllowedChannels(guildId, commandName) {
  const config = loadServerConfig(guildId);
  return config?.channels?.[commandName] || [];
}

// ==================== GESTION RÔLES ====================

function addRolePermission(guildId, permissionType, roleId) {
  const config = loadServerConfig(guildId);
  if (!config) return false;
  if (!config.roles[permissionType]) config.roles[permissionType] = [];
  if (!config.roles[permissionType].includes(String(roleId))) {
    config.roles[permissionType].push(String(roleId));
    saveServerConfig(guildId, config);
    return true;
  }
  return false;
}

function removeRolePermission(guildId, permissionType, roleId) {
  const config = loadServerConfig(guildId);
  if (!config) return false;
  const idx = (config.roles[permissionType] || []).indexOf(String(roleId));
  if (idx !== -1) {
    config.roles[permissionType].splice(idx, 1);
    saveServerConfig(guildId, config);
    return true;
  }
  return false;
}

function getAllowedRoles(guildId, permissionType) {
  const config = loadServerConfig(guildId);
  return config?.roles?.[permissionType] || [];
}

// ==================== SALONS SANS COINS ====================

function getNoCoinsChannels(guildId) {
  const config = loadServerConfig(guildId);
  return config?.no_coins_channels || [];
}

function addNoCoinsChannel(guildId, channelId) {
  const config = loadServerConfig(guildId);
  if (!config) return false;
  if (!config.no_coins_channels) config.no_coins_channels = [];
  if (!config.no_coins_channels.includes(String(channelId))) {
    config.no_coins_channels.push(String(channelId));
    saveServerConfig(guildId, config);
    return true;
  }
  return false;
}

function removeNoCoinsChannel(guildId, channelId) {
  const config = loadServerConfig(guildId);
  if (!config) return false;
  const idx = (config.no_coins_channels || []).indexOf(String(channelId));
  if (idx !== -1) {
    config.no_coins_channels.splice(idx, 1);
    saveServerConfig(guildId, config);
    return true;
  }
  return false;
}

function isCoinsDisabledChannel(guildId, channelId) {
  return getNoCoinsChannels(guildId).includes(String(channelId));
}

module.exports = {
  OWNER_ID,
  initServerConfig,   // ré-exporté pour rétrocompatibilité
  loadServerConfig,   // ré-exporté pour rétrocompatibilité
  saveServerConfig,   // ré-exporté pour rétrocompatibilité
  checkChannelPermission,
  getAllowedChannel,
  checkRolePermission,
  addChannelPermission,
  removeChannelPermission,
  getAllowedChannels,
  addRolePermission,
  removeRolePermission,
  getAllowedRoles,
  getNoCoinsChannels,
  addNoCoinsChannel,
  removeNoCoinsChannel,
  isCoinsDisabledChannel,
};